File: ADT/dfs_bfs/1970.py
# Google, Meta


# DFS + Binary Search
class Solution:
    def __init__(self):
        self.ROW = 0
        self.COL = 0
        self.directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    def DFS(self, grid: List[List[int]], i: int, j: int) -> bool:
        if i < 0 or i >= self.ROW or j < 0 or j >= self.COL or grid[i][j] == 1:
            return False
        if i == self.ROW - 1:
            return True
        grid[i][j] = 1
        for di, dj in self.directions:
            new_i, new_j = i + di, j + dj
            if self.DFS(grid, new_i, new_j):
                return True
        return False

    def canCross(self, cells: List[List[int]], mid: int) -> bool:
        grid = [[0] * self.COL for _ in range(self.ROW)]
        for i in range(mid + 1):
            x, y = cells[i][0] - 1, cells[i][1] - 1
            grid[x][y] = 1
        for j in range(self.COL):
            if grid[0][j] == 0 and self.DFS(grid, 0, j):
                return True
        return False

    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        self.ROW = row
        self.COL = col
        l, r = 0, len(cells) - 1
        lastDay = 0
        while l <= r:
            mid = l + (r - l) // 2
            if self.canCross(cells, mid):
                lastDay = mid + 1
                l = mid + 1
            else:
                r = mid - 1
        return lastDay


# BFS + Binary Search
from collections import deque


class Solution:
    def __init__(self):
        self.ROW = 0
        self.COL = 0
        self.directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    def BFS(self, grid: List[List[int]], i: int, j: int) -> bool:
        que = deque()
        que.append((i, j))
        grid[i][j] = 1
        while que:
            x, y = que.popleft()
            if x == self.ROW - 1:
                return True
            for dir_x, dir_y in self.directions:
                new_x = x + dir_x
                new_y = y + dir_y
                if (
                    0 <= new_x < self.ROW
                    and 0 <= new_y < self.COL
                    and grid[new_x][new_y] == 0
                ):
                    que.append((new_x, new_y))
                    grid[new_x][new_y] = 1
        return False

    def canCross(self, cells: List[List[int]], mid: int) -> bool:
        grid = [[0] * self.COL for _ in range(self.ROW)]
        for i in range(mid + 1):
            x, y = cells[i][0] - 1, cells[i][1] - 1
            grid[x][y] = 1
        for j in range(self.COL):
            if grid[0][j] == 0 and self.BFS(grid, 0, j):
                return True
        return False

    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        self.ROW = row
        self.COL = col
        l, r = 0, len(cells) - 1
        lastDay = 0
        while l <= r:
            mid = l + (r - l) // 2
            if self.canCross(cells, mid):
                lastDay = mid + 1
                l = mid + 1
            else:
                r = mid - 1
        return lastDay

File: ADT/dfs_bfs/1970.cpp
// Google, Meta

// DFS + Binary Search
// TC = O{(row * col) * logn} [logn for binary search]
class Solution {
public:
    int ROW;
    int COL;
    vector<vector<int>> directions{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    bool DFS(vector<vector<int>>& grid, int i, int j) {
        if (i < 0 || i >= ROW || j < 0 || j >= COL || grid[i][j] == 1) {
            return false;
        }
        if (i == ROW - 1) {
            return true;
        }
        grid[i][j] = 1;
        for (vector<int>& dir : directions) {
            int new_i = i + dir[0];
            int new_j = j + dir[1];
            if (DFS(grid, new_i, new_j))
                return true;
        }
        return false;
    }
    bool canCross(vector<vector<int>>& cells, int mid) {
        vector<vector<int>> grid(ROW, vector<int>(COL));
        for (int i = 0; i <= mid; i++) {
            int x = cells[i][0] - 1;
            int y = cells[i][1] - 1;
            grid[x][y] = 1;
        }
        for (int j = 0; j < COL; j++) {
            if (grid[0][j] == 0 && DFS(grid, 0, j)) {
                return true;
            }
        }
        return false;
    }
    int latestDayToCross(int row, int col, vector<vector<int>>& cells) {
        ROW = row;
        COL = col;
        int l = 0;
        int r = cells.size() - 1;
        int lastDay = 0;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (canCross(cells, mid)) {
                lastDay = mid + 1;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return lastDay;
    }
};



// BFS + Binary Search
class Solution {
public:
    int ROW;
    int COL;
    vector<vector<int>> directions{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    bool BFS(vector<vector<int>>& grid, int i, int j) {
        queue<pair<int, int>> que;
        que.push({i, j});
        grid[i][j] = 1;
        while (!que.empty()) {
            auto temp = que.front();
            que.pop();
            int x = temp.first;
            int y = temp.second;
            if (x == ROW - 1)
                return true;
            for (vector<int>& dir : directions) {
                int new_x = x + dir[0];
                int new_y = y + dir[1];
                if (new_x >= 0 && new_x < ROW && new_y >= 0 && new_y < COL &&
                    grid[new_x][new_y] == 0) {
                    que.push({new_x, new_y});
                    grid[new_x][new_y] = true;
                }
            }
        }
        return false;
    }
    bool canCross(vector<vector<int>>& cells, int mid) {
        vector<vector<int>> grid(ROW, vector<int>(COL));
        for (int i = 0; i <= mid; i++) {
            int x = cells[i][0] - 1;
            int y = cells[i][1] - 1;
            grid[x][y] = 1;
        }
        for (int j = 0; j < COL; j++) {
            if (grid[0][j] == 0 && BFS(grid, 0, j)) {
                return true;
            }
        }
        return false;
    }
    int latestDayToCross(int row, int col, vector<vector<int>>& cells) {
        ROW = row;
        COL = col;
        int l = 0;
        int r = cells.size() - 1;
        int lastDay = 0;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (canCross(cells, mid)) {
                lastDay = mid + 1;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return lastDay;
    }
};